sql: properly log copy error on error

otan opened the pull request:
This commit fixes an earlier commit by ensuring maybeLogStatement for
copy actually logs the error that comes out of it.

Release justification: bug fix
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86870)
<!-- Reviewable:end -->

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-:-NAKWDM_2fhE-RzNxK8D:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[`pkg/sql/conn_executor.go` line 2459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAKW0Q-ANkd3U8339-5:-NAKW0Q-ANkd3U8339-6:blyage9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c6c84e5938fd7e9b3aaa5ca4d70bced7a6cf8b/pkg/sql/conn_executor.go#L2459)):*
> ```Go
> 			ex.extraTxnState.txnCounter,
> 			cm.numInsertedRows(),
> 			copyError,
> ```

Man I actually looked at this retErr thing and thought about if it was correct and convinced myself it was :facepalm: My bad!


<!-- Sent from Reviewable.io -->

otan commented:
probably worth another look, added some extra state to set to make it not panic!

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86870)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @otan, and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 2424 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ8GlXAaoP2xsIP-fR:-NAQ8GlXAaoP2xsIP-fS:bknc5ht) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2424)):*
> ```Go
> 	ann := tree.MakeAnnotations(0)
> 	ex.planner.extendedEvalCtx.Context.Annotations = &ann
> 	ex.planner.extendedEvalCtx.Context.Placeholders = &tree.PlaceholderInfo{}
> ```

i don't follow the reasoning for zeroing out annotations and placeholders or changing ex.planner.stmt
___
*[`pkg/sql/conn_executor.go` line 2427 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ7PhZ4YXbsWfglXPj:-NAQ7Ph_-1KUmHmfrDbp:b-t0of48) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2427)):*
> ```Go
> 	ex.planner.curPlan.stmt = &ex.planner.stmt
> 	var cm copyMachineInterface
> 	var err error
> ```

i thought it was more readable in the earlier revision where it was named `coopyErr`
___
*[`pkg/sql/conn_executor.go` line 2430 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ7Dbp3ccsraEywEvG:-NAQ7Dbp3ccsraEywEvH:bjot0sy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2430)):*
> ```Go
> 	defer func() {
> 		var numInsertedRows int
> 		if cm != nil {
> ```

you don't need this nil check, since you made the method able to handle nil receivers. i'd say do one nil check or the other
___
*[`pkg/sql/conn_executor.go` line 2434 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ85rZ0-gDn9Vf_Wek:-NAQ85rZ0-gDn9Vf_Wel:b-y9a3pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2434)):*
> ```Go
> 		}
> 		// These fields are not available in COPY, so use the empty value.
> 		var stmtFingerprintID roachpb.StmtFingerprintID
> ```

can we call `roachpb.ConstructStatementFingerprintID` instead of using the empty value?
___
*[`pkg/sql/event_log.go` line 279 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ8TOb52MR1HDV2Bgq:-NAQ8TOb52MR1HDV2Bgr:b20l5t0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/event_log.go#L279)):*
> ```Go
> 	// Inject the common fields into the payload provided by the caller.
> 	injectCommonFields := func(event logpb.EventPayload) error {
> 		if txn != nil {
> ```

in which cases is txn nil?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86870)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @otan, and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 2459 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAKW0Q-ANkd3U8339-5:-NARymAP-z-fOhjHTgjX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5c6c84e5938fd7e9b3aaa5ca4d70bced7a6cf8b/pkg/sql/conn_executor.go#L2459)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Man I actually looked at this retErr thing and thought about if it was correct and convinced myself it was :facepalm: My bad!
</blockquote></details>

Done.
___
*[`pkg/sql/conn_executor.go` line 2424 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ8GlXAaoP2xsIP-fR:-NARyWER88TGiPC9QlPj:b1it7xa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2424)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't follow the reasoning for zeroing out annotations and placeholders or changing ex.planner.stmt
</blockquote></details>

without this it panics on https://github.com/cockroachdb/cockroach/blob/92b550b6c4ec0060bfc7b3eddbc9c0950334e8dd/pkg/sql/event_log.go#L189
___
*[`pkg/sql/conn_executor.go` line 2427 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ7PhZ4YXbsWfglXPj:-NARyeS897_bp86rr8zk:b-kby030) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2427)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i thought it was more readable in the earlier revision where it was named `coopyErr`
</blockquote></details>

i'm 50:50, this is the original error
___
*[`pkg/sql/conn_executor.go` line 2430 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ7Dbp3ccsraEywEvG:-NARygFqCQ5DgmW0W2PB:b23slb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/conn_executor.go#L2430)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

you don't need this nil check, since you made the method able to handle nil receivers. i'd say do one nil check or the other
</blockquote></details>

no you do, because the interface could point to nil if `cm` never gets registered.
___
*[`pkg/sql/event_log.go` line 279 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAQ8TOb52MR1HDV2Bgq:-NARymmo21Ung1dHYegA:b6ydmpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b91a2da6ef6490d0fad411cdb04ddd7a294f72c2/pkg/sql/event_log.go#L279)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

in which cases is txn nil?
</blockquote></details>

when copy is the first statement. we never actually set `txn` up in the `planner` for copy.


<!-- Sent from Reviewable.io -->

otan commented:
@rafiss mind giving this another look? having to rebase a lot so fearing the backport even more now 🙃

rafiss reviewed (approved):
this needs a small change to the FmtCtx, then feel free to merge

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-:-NAfRfYlC9lVbjhAEzms:br33gjq)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @otan, and @rafiss)

___
*[`pkg/sql/conn_executor.go` line 2448 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/86870#-NAfQnyKAHX2SZv5rr5I:-NAfQnyLBUuZLVFH7f-m:bu48wz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2f54617a0681ca6f1cd46b0e3b0c36d037d8463/pkg/sql/conn_executor.go#L2448)):*
> ```Go
> 		}
> 		// These fields are not available in COPY, so use the empty value.
> 		f := tree.NewFmtCtx(tree.FmtAnonymize | tree.FmtHideConstants)
> ```

for making the fingerprint, we only want FmtHideConstants. we don't want FmtAnonymize, since `COPY t1 FROM STDIN` should not have the same fingerprint as `COPY t2 FROM STDIN`

part of the confusion here is probably that the param of `ConstructStatementFingerprintID` is named `anonymizedStmt`. I missed that when i made 1944c522001ee57de7c419397db7993e14fc0e95


<!-- Sent from Reviewable.io -->

otan committed:
sql: properly log copy error on error

This commit fixes an earlier commit by ensuring maybeLogStatement for
copy actually logs the error that comes out of it.

Release justification: bug fix
Release note: None

otan commented:
bors r=rafiss,cucaroach
